File: app/src/main/java/com/kyilmaz/neurocomet/SecurityUtils.kt
package com.kyilmaz.neurocomet

import android.util.Base64

/**
 * Utility for basic string obfuscation to prevent simple string searches in the APK.
 * Note: This is not "unbreakable" encryption, but it's much better than plain text
 * for discouraging casual bad actors.
 */
object SecurityUtils {

    // Simple XOR key - should be different from the one used in build.gradle.kts if possible,
    // but for simplicity we'll keep them consistent.
    private const val XOR_KEY = "neurocomet_internal_security_key_2025"

    /**
     * De-obfuscates a string that was obfuscated with XOR and Base64 encoded.
     */
    fun decrypt(obfuscated: String?): String {
        if (obfuscated == null || obfuscated.isEmpty() || obfuscated == "null" || obfuscated == "\"\"") {
            return ""
        }

        return try {
            val decoded = Base64.decode(obfuscated, Base64.DEFAULT)
            val result = StringBuilder()
            for (i in decoded.indices) {
                result.append((decoded[i].toInt() xor XOR_KEY[i % XOR_KEY.length].toInt()).toChar())
            }
            result.toString()
        } catch (e: Exception) {
            ""
        }
    }

    /**
     * Obfuscates a string using XOR and Base64 encoding.
     * Use this to generate obfuscated strings for your local.properties if you want
     * an extra layer of security, though build.gradle.kts will handle it automatically now.
     */
    fun encrypt(plain: String): String {
        val bytes = plain.toByteArray(Charsets.UTF_8)
        val obfuscated = ByteArray(bytes.size)
        for (i in bytes.indices) {
            obfuscated[i] = (bytes[i].toInt() xor XOR_KEY[i % XOR_KEY.length].toInt()).toByte()
        }
        return Base64.encodeToString(obfuscated, Base64.NO_WRAP)
    }
}

File: flutter_app/lib/core/utils/security_utils.dart
import 'dart:convert';

/// Utility for basic string obfuscation to prevent simple string searches in the binary.
/// Matches the implementation in the native Android version.
class SecurityUtils {
  SecurityUtils._();

  // Simple XOR key - MUST match the one in build.gradle.kts
  static const String _xorKey = 'neurocomet_internal_security_key_2025';

  /// De-obfuscates a string that was obfuscated with XOR and Hex encoded.
  static String decrypt(String? obfuscated) {
    if (obfuscated == null ||
        obfuscated.isEmpty ||
        obfuscated == 'null' ||
        obfuscated == '""') {
      return '';
    }

    try {
      final result = StringBuffer();
      for (int i = 0; i < obfuscated.length; i += 2) {
        final hex = obfuscated.substring(i, i + 2);
        final byte = int.parse(hex, radix: 16);
        final keyChar = _xorKey[(i ~/ 2) % _xorKey.length].codeUnitAt(0);
        result.writeCharCode(byte ^ keyChar);
      }
      return result.toString();
    } catch (e) {
      return '';
    }
  }

  /// Obfuscates a string using XOR and Hex encoding.
  static String encrypt(String plain) {
    final bytes = utf8.encode(plain);
    final result = StringBuffer();
    for (int i = 0; i < bytes.length; i++) {
      final obfuscatedByte = bytes[i] ^ _xorKey[i % _xorKey.length].codeUnitAt(0);
      result.write((obfuscatedByte & 0xFF).toRadixString(16).padLeft(2, '0'));
    }
    return result.toString();
  }
}
